File: cs/days/day01.cs
namespace Shunty.AoC;

// https://adventofcode.com/2023/day/1 - Trebuchet?!

public class Day01 : AocDaySolver
{
    public static int DayNumber() => 1;

    public async Task Solve()
    {
        var input = await AocUtils.GetDayLines(DayNumber());

        // Part 1
        var sum = 0;
        foreach (var s in input)
        {
            var n = s.Where(c => c >= '0' && c <= '9');
            sum += ((int)(n.First() - '0') * 10) + (int)(n.Last() - '0');
        }
        this.ShowDayResult(1, sum);

        // Part 2
        var digitmap = new Dictionary<string, int>
        {
            {"0", 0},{"1", 1},{"2", 2},{"3", 3},{"4", 4},{"5", 5},{"6", 6},{"7", 7},{"8", 8},{"9", 9},
            {"one", 1}, {"two", 2}, {"three", 3}, {"four", 4}, {"five", 5},
            {"six", 6}, {"seven", 7}, {"eight", 8}, {"nine", 9},
        };
        var sum2 = 0;
        foreach (var s in input)
        {
            var first = -1;
            var last = 0;
            for (var index = 0; index < s.Length; index++)
            {
                foreach (var (k,v) in digitmap)
                {
                    if (s.AsSpan(index).StartsWith(k))
                    {
                        if (first == -1)
                            first = v * 10;
                        last = v;
                        // Don't do this! What about "...nineight..." or "...threeightwo.." types of sequences
                        //index += k.Length - 1;
                        if (k.Length > 1)
                            index += k.Length - 2;
                        break;
                    }
                }
            }
            sum2 += first + last;
        }
        this.ShowDayResult(2, sum2);
    }
}

File: py/day01.py
import os

# https://adventofcode.com/2023/day/1

with open(os.path.dirname(os.path.realpath(__file__)) + "/../input/day01-input", "r") as f:
    input = [line.strip() for line in f.readlines()]

# Part 1
sum1 = 0
for s in input:
    nums = [int(c) for c in s if c >= '0' and c <= '9']
    sum1 += (nums[0] * 10) + nums[-1]

print("Part 1:", sum1)

# Part 2
digits = ["one", "two", "three", "four", "five", "six", "seven", "eight", "nine", "0", "1", "2", "3", "4", "5", "6", "7", "8", "9"]
sum2 = 0
for s in input:
    i = 0
    first = -1
    last = -1
    # Find first, scan forward
    while i < len(s):
        for idx, digit in enumerate(digits):
            if s[i:].startswith(digit):
                first = 10 * (idx + 1 if len(digit) > 1 else int(digit))
                break
        if first > 0:
            break
        i += 1

    # Find last, scan backward
    i = len(s) - 1
    while i >= 0:
        for idx, digit in enumerate(digits):
            if s[i:].startswith(digit):
                last = idx + 1 if len(digit) > 1 else int(digit)
                break
        if last >= 0:
            break
        i -= 1

    sum2 += first + last

print("Part 2:", sum2)
